File: benchmark/cases/binary_trees/binary_trees.py
import os
import time


class Node:
    __slots__ = ("l", "r")

    def __init__(self, l, r):
        self.l = l
        self.r = r


def make(d):
    if d == 0:
        return Node(None, None)
    return Node(make(d - 1), make(d - 1))


def check(n):
    if n.l is None:
        return 1
    return 1 + check(n.l) + check(n.r)


depth = int(os.environ["BENCH_SCALE"])
t0 = time.perf_counter_ns()
root = make(depth)
count = check(root)
t1 = time.perf_counter_ns()
print("bintree: nodes =", count)
print("__bench_ns:", t1 - t0)

File: benchmark/cases/binary_trees/binary_trees.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct Node {
    struct Node *l, *r;
} Node;

Node *make(int d) {
    Node *n = malloc(sizeof(Node));
    if (d == 0) {
        n->l = NULL;
        n->r = NULL;
    } else {
        n->l = make(d - 1);
        n->r = make(d - 1);
    }
    return n;
}

long check(Node *n) {
    if (n->l == NULL)
        return 1;
    return 1 + check(n->l) + check(n->r);
}

int main(void) {
    int depth = atoi(getenv("BENCH_SCALE"));
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    Node *root = make(depth);
    long count = check(root);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    long ns = (t1.tv_sec - t0.tv_sec) * 1000000000L + (t1.tv_nsec - t0.tv_nsec);
    printf("bintree: nodes = %ld\n", count);
    printf("__bench_ns: %ld\n", ns);
    return 0;
}
